src/platform/cc13x2_26x2/OTAImageProcessorImpl.cpp

```diff
@@ -0,0 +1,436 @@
+}
+
+void OTAImageProcessorImpl::HandleFinalize(intptr_t context)
+{
+    ExtImageInfo_t header;
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        return;
+    }
+
+    if (!readExtFlashImgHeader(imageProcessor->mNvsHandle, &(header.fixedHdr)))
+    {
+        return;
+    }
+    header.extFlAddr = IMG_START;
+    header.counter   = 0x0;
+
+    if (validateExtFlashImage(imageProcessor->mNvsHandle))
+    {
+        // only write the meta header if the crc check passes
+        writeExtFlashMetaHeader(imageProcessor->mNvsHandle, &header);
+    }
+    else
+    {
+        // ensure the external image is not mistaken for a valid image
+        eraseExtFlashMetaHeader(imageProcessor->mNvsHandle);
+    }
+
+    imageProcessor->ReleaseBlock();
+
+    ChipLogProgress(SoftwareUpdate, "OTA image downloaded");
+}
+
+void OTAImageProcessorImpl::HandleApply(intptr_t context)
+{
+    ExtImageInfo_t header;
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        return;
+    }
+
+    if (!readExtFlashMetaHeader(imageProcessor->mNvsHandle, &header))
+    {
+        return;
+    }
+    header.fixedHdr.imgCpStat = NEED_COPY;
+
+    writeExtFlashMetaHeader(imageProcessor->mNvsHandle, &header);
+
+    // reset SoC to kick BIM
+    SysCtrlSystemReset();
+}
+
+void OTAImageProcessorImpl::HandleAbort(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        return;
+    }
+
+    if (!eraseExtFlashMetaHeader(imageProcessor->mNvsHandle))
+    {
+        imageProcessor->mDownloader->OnPreparedForDownload(CHIP_ERROR_WRITE_FAILED);
+    }
+
+    NVS_close(imageProcessor->mNvsHandle);
+    imageProcessor->ReleaseBlock();
+}
+
+void OTAImageProcessorImpl::HandleProcessBlock(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    else if (imageProcessor->mDownloader == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "mDownloader is null");
+        return;
+    }
+
+    // TODO: Process block header if any
+
+    if (!writeExtFlashImgPages(imageProcessor->mNvsHandle, imageProcessor->mParams.downloadedBytes, imageProcessor->mBlock))
+    {
+        imageProcessor->mDownloader->EndDownload(CHIP_ERROR_WRITE_FAILED);
+        return;
+    }
+
+    imageProcessor->mParams.downloadedBytes += imageProcessor->mBlock.size();
+    imageProcessor->mDownloader->FetchNextData();
+}
+
+CHIP_ERROR OTAImageProcessorImpl::SetBlock(ByteSpan & block)
+{
+    if ((block.data() == nullptr) || block.empty())
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    // Allocate memory for block data if it has not been done yet
+    if (mBlock.empty())
```

Comment:
This is buggy.  Please copy one of the non-buggy implementations of SetBlock instead...

@selissia